packages/js/product-editor/src/components/header/hooks/use-preview/use-preview.tsx

```diff
@@ -0,0 +1,124 @@
+/**
+ * External dependencies
+ */
+import { Product } from '@woocommerce/data';
+import { Button } from '@wordpress/components';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { MouseEvent } from 'react';
+
+export function usePreview( {
+	productId,
+	disabled,
+	onClick,
+	onSaveSuccess,
+	onSaveError,
+	...props
+}: Omit< Button.AnchorProps, 'aria-disabled' | 'variant' | 'href' > & {
+	productId: number;
+	onSaveSuccess?( product: Product ): void;
+	onSaveError?( error: Error ): void;
+} ): Button.AnchorProps {
+	const anchorRef = useRef< HTMLAnchorElement >();
+
+	const { permalink, productStatus, hasEdits } = useSelect(
+		( select ) => {
+			const { getEditedEntityRecord, hasEditsForEntityRecord } =
+				select( 'core' );
+
+			const product = getEditedEntityRecord< Product | undefined >(
+				'postType',
+				'product',
+				productId
+			);
+
+			return {
+				permalink: product?.permalink,
+				productStatus: product?.status,
+				hasEdits: hasEditsForEntityRecord< boolean >(
+					'postType',
+					'product',
+					productId
+				),
+			};
+		},
+		[ productId ]
+	);
+
+	const { editEntityRecord, saveEditedEntityRecord } = useDispatch( 'core' );
+
+	let previewLink: string | undefined;
+	if ( typeof permalink === 'string' ) {
+		if ( permalink.includes( '?' ) ) {
```

Comment:
I think this probably works okay, but just to prevent any weird edge cases we could use the browser API:

```
const url = new URL( permalink );
url.searchParams.append( 'preview', 'true' );
```